### ISSUE TITLE
The backend differentiates functions when it is not entitled to

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8248]
Consider the following test models
```
package TestSmoothOrder
  model M1
    function f
      input Real x;
      output Real y;
    algorithm
      if x > 0 then 
        y := x^2 +1;
      else
        y := -x^2;
      end if;
    end f;
    Real x, y;
  equation
    x = f(time-0.5);
    der(x) = y;
  end M1;
  
  model M2
    function f
      input Real x;
      output Real y;
    algorithm
      if x > 0 then 
        y := x^2;
      else
        y := -x^2;
      end if;
    end f;
    Real x, y;
  equation
    x = f(time-0.5);
    der(x) = y;
  end M2;
  
  model M3
    function f
      input Real x;
      output Real y;
    algorithm
      if x > 0 then 
        y := x^2;
      else
        y := -x^2;
      end if;
    annotation(smoothOrder = 1);
    end f;
    Real x, y;
  equation
    x = f(time-0.5);
    der(x) = y;
  end M3;
end TestSmoothOrder;
```
All three models currently simulate with OMC.

The three models involve functions that include if-statements, so they are potentially discontinuous. The compiler is not smart enough to figure out if they are smooth enough across the switching of the condition. Hence, it should not even try to differentiate function `f`, unless `f` is explicitly declared as being smooth enough to do so.

Bottom line: only M3 should simulate, while M1 and M2 should abort with error message "Cannot differentiate function f" or something like that. Continuing with the compilation in those cases is dangerous, because if the function is _actually_ non-differentiable, as in the case of M1, the consequences on the results can be dire.

**Comment by casella** [comment]
See #14698 for a further example by @MartinOtter including external functions.

@kabdelhak, @phannebohm we should really fix this. Can you consider that for next year?

**Comment by casella** [comment]
See also #8658 and #12912

**Comment by casella** [comment]
Keeping @bernhardbachmann in the loop
